mixnet/mixnet.py

```diff
@@ -0,0 +1,66 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import List, Self, TypeAlias
+
+from cryptography.hazmat.primitives.asymmetric.x25519 import (X25519PrivateKey,
+                                                              X25519PublicKey)
+
+from mixnet.bls import BlsPrivateKey, BlsPublicKey
+from mixnet.fisheryates import FisherYates
+
+NodeId: TypeAlias = BlsPublicKey
+# 32-byte that represents an IP address and a port of a mix node.
+NodeAddress: TypeAlias = bytes
+
+
+@dataclass
+class Mixnet:
+    mix_nodes: List[MixNode]
+
+    # Build a new topology deterministically using an entropy.
+    # The entropy is expected to be injected from outside.
+    #
+    # TODO: Implement constructing a new topology in advance to minimize the topology transition time.
+    #       https://www.notion.so/Mixnet-Specification-807b624444a54a4b88afa1cc80e100c2?pvs=4#9a7f6089e210454bb11fe1c10fceff68
+    def build_topology(
+        self,
+        entropy: bytes,
+        n_layers: int,
+        n_nodes_per_layer: int,
+    ) -> MixnetTopology:
+        assert n_nodes_per_layer * n_layers < len(self.mix_nodes)
+
+        shuffled = FisherYates.shuffle(self.mix_nodes, entropy)
+        sampled = shuffled[: n_nodes_per_layer * n_layers]
+        layers = []
+        for l in range(n_layers):
+            start = l * n_nodes_per_layer
+            layer = sampled[start : start + n_nodes_per_layer]
+            layers.append(layer)
```

Comment:
I have the feeling this could be a `pick_layer` or `sample_layer` method.

Reply:
I just tried to extract this as a function, but the function was not neat as I expected because the function needs an assumption that the length of array must be always `n_nodes_per_layer * n_layer`.
If the implementation above is not very complicated, I think it's better to have it because it shows the full context about how many nodes were sampled.